### PR TITLE
[WIP] Move ocaml_migrate_parsetree ppx driver fucntionality from ppx_deriving_main to ppx_deriving module.

### LINKED ISSUE #134: Instruction on using ppx_deriving with jbuilder?
Hi,
I am attempting to use the latest github version of `ppx_deriving` with jbuilder. However, it doesn't seem to be working with the stanza `(preprocess (pps (ppx_deriving.std)))`. Could you please provide some guidance on how one can use `ppx_deriving` with `jbuilder`.

Thanks.

### PR DESCRIPTION
**Pull request by bikalgurung**
@whitequark fixes #134.

I will test this with a jbuilder project. After this is complete, I will remove the [WIP] and it should be ready to merge.

### CODE REVIEW

**Comment by whitequark** [comment]
This is tested, correct?

File: src/ppx_deriving_main.ml [hunk 1]
```diff
@@ -1,92 +1,4 @@
...
-      plugins_to_load := [];
-      mapper.Ast_mapper.structure mapper rest
-    | items -> Ppx_deriving.mapper.Ast_mapper.structure mapper items in
-  { Ppx_deriving.mapper with Ast_mapper.structure }
-
-let () =
-  Driver.register ~name:"ppx_deriving" ~args:args_spec
-    Versions.ocaml_405 rewriter;
-  Driver.run_main ()
+include Ppx_deriving
```
Comment (rgrinberg) [inline src/ppx_deriving_main.ml]: Why include and not open here?

Comment (bikalgurung) [inline src/ppx_deriving_main.ml]: yep. good catch! I will change this to open since we are not re-exporting `Ppx_deriving` here.

**Comment by bikalgurung** [comment]
@whitequark I haven't had much success with testing yet. ppx_deriving is still not picked up yet. The standalone version doesn't seem to be working too. It just hangs and never terminates. Reviewing/testing further. Will report on issues later.

**Comment by bikalgurung** [comment]
@diml I have moved the ocaml-migrate-parsetree functionality to `Ppx_deriving` from Ppx_deriving_mainl`. However, this doesn't seem to work - with both jbuilder and standalone version of `ppx_deriving` executable. I executed the stand-alone `ppx_deriving` and it just hangs, i.e. doesn't terminate. Not sure what I am missing here. Could you please help?

**Comment by diml** [comment]
@bikalgurung you need to build the ppx_deriving library with `linkall`, to be sure it gets linked

**Comment by diml** [comment]
Actually, ppx_deriving shouldn't need to be linked with `-linkall`, just the plugings. @bikalgurung can you give an example of something you tried that didn't work?

**Comment by bikalgurung** [comment]
@diml below is what I am using to test.
```
$ cat jbuild
(jbuild_version 1)

(executables
  ((names (foo))
    (libraries (ppx_deriving.runtime))
    (preprocess (pps (ppx_deriving.show)))))
```

```
$ cat foo.ml
type foo = A of int | B of float
[@@deriving show]

let () =
  let f = A 100 in 
  (* show_foo f; *)
  print_string "...";
```
Attempting with standalone version.
```
$ _build/.ppx/default/ppx_deriving.show/ppx.exe foo.ml 
type foo =
  | A of int 
  | B of float [@@deriving show]
let () = let f = A 100  in print_string "..."
```
Trying with `ppx_deriving` standalone executable.
```
$ `ocamlfind query ppx_deriving`/ppx_deriving -deriving-plugin `ocamlfind query ppx_deriving`/ppx_deriving_show.cma foo.ml

```
The last one doesn't terminate.

**Comment by diml** [comment]
For the first problem, try adding this to `_tags`:

```
<src_plugins/*>: linkall
```

For the second one, try running the command in gdb, type Ctrl+C and then `bt` to get a backtrace. Hopefully you'll get an idea where it's stuck.

**Comment by bikalgurung** [comment]
@diml @whitequark 
I tried `<src_plugins/*>: linkall` in `ppx_deriving`. But now I am getting an error while trying to build my test jbuild project. 

Executing `$ jbuilder build foo.exe` gives the following error. 
```
File "_none_", line 1:
Error: No implementations provided for the following modules:
         Findlib referenced from /home/bg/.opam/4.04.1+flambda/lib/ppx_deriving/ppx_deriving.cmxa(Ppx_deriving)
         Fl_dynload referenced from /home/bg/.opam/4.04.1+flambda/lib/ppx_deriving/ppx_deriving.cmxa(Ppx_deriving)
```
I tried adding `findlib.dynload` to executable dependency. 
```
$ cat jbuild
(jbuild_version 1)

(executables
  ((names (foo))
    (libraries (findlib.dynload))    
    (preprocess (pps (ppx_deriving.show)))))
```
But jbuilder doesn't seem to pick up the dependency. Here is the significant bits of `_build/log` file
```
$ (cd _build/default && /home/bg/.opam/4.04.1+flambda/bin/ocamlopt.opt -o ../.ppx/default/ppx_deriving.show/ppx.exe -I /home/bg/.opam/4.04.1+flambda/lib/ocaml -I /home/bg/.opam/4.04.1+flambda/lib/ocaml-migrate-parsetree -I /home/bg/.opam/4.04.1+flambda/lib/ocaml-migrate-parsetree/driver-main -I /home/bg/.opam/4.04.1+flambda/lib/ocaml/compiler-libs -I /home/bg/.opam/4.04.1+flambda/lib/ppx_deriving -I /home/bg/.opam/4.04.1+flambda/lib/ppx_tools_versioned -I /home/bg/.opam/4.04.1+flambda/lib/result ocamlcommon.cmxa migrate_parsetree.cmxa dynlink.cmxa ppx_tools_versioned.cmxa result.cmxa ppx_deriving.cmxa ppx_deriving_show.cmxa migrate_parsetree_driver_main.cmxa)
> File "_none_", line 1:
> Error: No implementations provided for the following modules:
>          Findlib referenced from /home/bg/.opam/4.04.1+flambda/lib/ppx_deriving/ppx_deriving.cmxa(Ppx_deriving)
>          Fl_dynload referenced from /home/bg/.opam/4.04.1+flambda/lib/ppx_deriving/ppx_deriving.cmxa(Ppx_deriving)
[2]
```
Doing `$ ocamlobjinfo _build/src/ppx_deriving.cmxa` gives the following output. 
```
File _build/src/ppx_deriving.cmxa
Extra C object files:
Extra C options:
Name: Ppx_deriving
CRC of implementation: 13ebbdb94a298a052f42e88904678f9e
Globals defined:
	Ppx_deriving
Interfaces imported:
	94a8285185889dd7e08dcacd9866b16e	Warnings
	6fb70405f4f231a940f7911af46504e5	Sys
	bc6375cfdeb96595c8a1276bd08c1fb2	Syntaxerr
	609a5859ae42eeafa6c2b0f8f5e2ee85	String
	cf3455f46534ce0e93f6aaf609c4197b	Set
	8d520e04228ca206985b0b85d46feb62	Result
	62d860c471d4249640088bc945349d6b	Queue
	385453685a093b8637adddc0c7998065	Printf
	06dd368708a9b6e0a43b209946f719bf	Ppx_tools_405
	f1b079a5ca10dab894310634c9bf0b2a	Ppx_deriving
	8a04747db2038b4afffe7ef54f63bd44	Pprintast
	eef96f967b03d53aceb35ab9ee61e6fc	Pervasives
	e48b13e961b0362c398e89bfdb7db3b7	Parsetree
	8cd714c1f9a08e7cb77aa18d52e830ed	Outcometree
	6adeb5fabc17b037c800e589fa54de9d	Nativeint
	9513c8e3cb1eef5b81d96eef1b3210b7	Misc
	5362216c3dd09678092c0abb04a03150	Migrate_parsetree_versions
	b3e46a0705d83b2930b9148e678b5d6e	Migrate_parsetree_driver
	62da2eb9f0db3f0cb44d1cd51a616e55	Migrate_parsetree
	79600cd97ccdb653e469726598ad3582	Map
	d3d7a366ba8a2fea49907aa72c755270	Longident
	43e66250e44bdfd6a21eb689c6060a7c	Location
	cd293869f2923980325b5f3354992c2d	List
	4d9c7eb04940c37e64ff967cfe13f973	Lexing
	7e97f2d9e2400938aba6fdcb2d54c302	Lazy
	b2b905603aad7b393017ff8889305fc8	Int64
	f4400e9e93b464d150ad42f31836792f	Int32
	4744e626acdfe40de6fc93fabc1a64f6	Hashtbl
	aed1d740331f56358bdc54ffe141c959	Format
	0271c675af8ae4b8cf8e13542537df0d	Fl_dynload
	daba487dd0aeb811a53438a953e3d9f5	Findlib
	b76309dbfd6df6e67384275ed924c982	Filename
	344030fecf95142b48f348779289130e	Dynlink
	b84048761577d28524b7d244313f8dd5	Digest
	2b18a802a1d42e9939b41f6b36f537c1	Char
	6c185d0cbcb037a5c9cdf06a0923af0d	CamlinternalLazy
	cbd5f2d6b649925222e1e9fb63b89db6	CamlinternalFormatBasics
	c932a66951c989158d32b4d60664f14e	Buffer
	09c75056122525994f6b902ba43cd59c	Asttypes
	e7ab6c986ff82f15b23c0a168e404a9c	Ast_mapper_class_405
	24497eb0d5c592b3a706da71ad7f7bb7	Ast_mapper
	86284440aea54c6596d76dc5769c2916	Ast_convenience_405
	afe4bd134a2e7811f8eca9c64dbbf0c2	Ast_405
	c3bf7bb1b59b36b9172586bd56cbbeee	Ast_404
	d7f967ea74fef4196cf9847f46826cdd	Ast_403
	ce6b93a2553675e8a2c3a4fd8667099d	Ast_402
	83626447aa49c1fc006c752026de61fb	Array
	f8b1bea86db5d3455940be2b75866274	Arg
Implementations imported:
	a8808e420cf4d96bb85a9a67724dc4c7	Ppx_tools_405
	469a5a32c6f236ff9a056547228a5853	Migrate_parsetree
	0c1af5f17c171f48cc87c27466e4c38e	CamlinternalFormat
	163c4f43df3e292b7a8d9fc92a087509	Ast_404
	187ab58a80a691b84849e06a850833c8	Random
	14ee2331a2095a4853dc9855b2ac9366	Bytes
	44304bd64df033811282052ccb4bb69f	Arg
	6a0692b6efdd6f1bd3ea6e80222575f5	Array
	6c4e6436b77da269d413eb0d2f6d546f	Sys
	3abff0d4498a36b80ebdf75ac8853731	Migrate_parsetree_driver
	--------------------------------	Fl_dynload
	5b193bb0dba7f00842b962d035041b7c	CamlinternalLazy
	--------------------------------	Findlib
	a1d562a75ee10e0c5153f67f511f3449	Dynlink
	c7af76f0607bd2d6e32a002828854e4a	Filename
	78123244b8a7a5c97a4350ac6fddc896	Set
	886432a3e1d3f5f62c8f7c6d59f163e6	Longident
	c321f120a5b82ccea74eb90ab5ca9b54	Ast_405
	f79a820a38ed31f708b374590a01a456	Ast_convenience_405
	a997e03e7925abd8c95540ef7a7da477	String
	86ee1b6be12e8471fbd29bcdfd3b96bd	List
	37fde9ee36ed21a0dd9a07698ed3ea0a	Pervasives
	a9f9b282d23cb2b2d30669670bcc1e43	Format
	5e0c5267964f0faf908e1854ce650795	Pprintast
	eb29eb970a691aa9511872ca564e5d52	Migrate_parsetree_versions
	486257e504d75e69f86310fc4443fdaa	Printf
	068524ed547a09d13e5130f98ea215d9	Location
	089a91ac6d342d06c1ca6403e819bbaa	Queue
	70e3551cb9bed4e543971f105ce17bee	Hashtbl
Flambda export information:
 approxs Globals: 
Symbols: 

functions
{ }

Currying functions: 5 9 4 6 3 2
Apply functions: 4 3 10 2
Send functions:
Force link: YES
```
From the output, I can see 
```
Implementation imported:
...
--------------------------------	Fl_dynload
5b193bb0dba7f00842b962d035041b7c	CamlinternalLazy
--------------------------------	Findlib
... 
```
Not sure if this is significant so I tried building `ppx_deriving` by adding  `<src/ppx_deriving.{ml,mli,byte,native}>: package(findlib.dynload), linkpkg` in `_tags` file. No luck.

What gives?

**Comment by diml** [comment]
The `----...` lines are because findlib doesn't install its cmx. This doesn't matter here, it just means that findlib functions can't be inlined.

Editing the _tags file will affect the build of ppx_deriving itself but not the users of ppx_deriving. You need to edit the `pkg/META` file for that.

BTW, intuitively I would think that all the dynlink machinery doesn't need to go in `Ppx_deriving`, since it provides functionalities that overlap with the use of a driver such as the ocaml-migrate-parsetree one.

FTR, I wrote [this branch](https://github.com/diml/ppx_deriving/commits/v4) which we currently use in the Jane Street development repository. It allows to use ppx_deriving with the omp driver and jbuilder.

**Comment by bikalgurung** [comment]
Phew!! This has been educational. Thanks @diml for your help and response.

**Comment by bikalgurung** [comment]
@whitequark it seems that the change needed to make the current version of `ppx_deriving` work with `jbuilder` is is a bit more elaborate than we envisioned earlier. @diml already has a version that works currently with `jbuilder`. Could we merge that instead? That version uses `ppx_tools` and I am guessing `ppx_tools` is much more stable than `ppx_tools_versioned`. I will close this pull request. 

@diml Thank you for your help. It's been educational.

**Pull request closed without merging**